app/concurrencylimiter.go

```diff
@@ -3,70 +3,155 @@ package app
 import (
 	"context"
 	"sync"
+
+	"github.com/pkg/errors"
 )
 
+var errQueueFull = errors.New("queue full")
+
 // concurrencyLimiter is a locking mechanism that allows setting a limit
 // on the number of simultaneous locks for a given arbitrary ID.
 type concurrencyLimiter struct {
-	max     int
-	count   map[string]int
-	pending map[string]chan struct{}
+	maxLock    int
+	maxQueue   int
+	lockCount  map[string]int
+	queueCount map[string]int
+	pending    map[string]*pendingList
 
 	mx sync.Mutex
 }
+type pendingReq struct {
+	lockCh     chan bool
+	prev, next *pendingReq
+}
+type pendingList struct{ head, tail *pendingReq }
+
+func (list *pendingList) newReq() *pendingReq {
+	req := &pendingReq{
+		lockCh: make(chan bool),
+		prev:   list.tail,
+	}
+
+	if list.head == nil {
+		list.head, list.tail = req, req
+	} else {
+		list.tail.next, list.tail = req, req
+	}
+
+	return req
+}
+func (list *pendingList) remove(req *pendingReq) bool {
+	if list.head == req {
+		list.head = req.next
+		return true
+	}
+
+	if list.tail == req {
+		list.tail = req.prev
+		if req.prev != nil {
+			req.prev.next = nil
+		}
+		return true
+	}
+
+	if req.next != nil {
+		req.next.prev = req.prev
+	}
+	if req.prev != nil {
+		req.prev.next = req.next
+		return true
+	}
```

Comment:
The above two sections only handles removing links for prev and next of `req` to `req`. Should we also handle removing same for `req` itself?
For example: List is `a = b = c`. And if we want to remove b from the list.
We shoud also say b.prev = nil  and b.next = nil. 
Meaning `req.prev = nil` and `req.next = nil`.

This way, we will remove all links from `b` to any other nodes. From my understanding, if there are no links, that would be equivalent to pointing to nil, But since links do exist here (since it's not a net new node), we should make sure they point to nil.

That way, in the future, if we ever introduce code involving `req.next` or `req.prev` it would not break.

Reply:
👍 actually uncovered a possible bug anyway. Added a test for the list specifically.